src/edu/csus/ecs/pc2/shadow/ShadowScoreboardComparator.java

```diff
@@ -0,0 +1,391 @@
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.PriorityQueue;
+
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import edu.csus.ecs.pc2.core.log.Log;
+import edu.csus.ecs.pc2.core.standings.json.ScoreboardJsonModel;
+import edu.csus.ecs.pc2.core.standings.json.Team;
+import edu.csus.ecs.pc2.core.standings.json.TeamScoreRow;
+import edu.csus.ecs.pc2.ui.ShadowCompareScoreboardPane.ScoreboardType;
+
+public class ShadowScoreboardComparator {
+    
+    private Log log ;
+    private ShadowController shadowController;
+    
+    /**
+     * Constructs a ShadowScoreboardComparator which can be used to generate comparison information for two
+     * scoreboards -- typically, a PC2 Shadow scoreboard and a Remote CCS scoreboard.
+     * 
+     * @param shadowController the {@link ShadowController} used by methods of this class; must not be null.
+     */
+    public ShadowScoreboardComparator (ShadowController shadowController) {
+        if (shadowController==null) {
+            throw new RuntimeException ("ShadowScoreboardComparator constructor called with null ShadowController");
+        }
+        this.shadowController = shadowController;
+        this.log = shadowController.getLog();
+    }
+
+    /**
+     * Returns an array of {@link ShadowScoreboardRowComparison}s, where each array element contains comparison 
+     * information between corresponding rows of the specified scoreboard JSON strings.
+     * 
+     *  The scoreboard JSON strings are expected to be in the format defined by the CLICS Contest API at
+     *  https://ccs-specs.icpc.io/contest_api#scoreboard.
+     *  
+     *  The returned array contains one {@link ShadowScoreboardRowComparison} object for each corresponding 
+     *  {@link TeamScoreRow} (that is, for each element of the "rows" array in the CLICS scoreboard JSON strings)
+     *  which is present in both scoreboards. If there is an element which is present in one of the JSON strings
+     *  but not in the other, the corresponding element in the returned array will contain a ScoreboardRowComparison 
+     *  object in which one of the TeamScoreRow entries is null.
+     *  
+     *  If an error occurs while converting either input JSON string to a collection of corresponding TeamScoreRow
+     *  objects, an array of size zero is returned.
+     *  
+     * @param pc2ScoreboardJson a CLICS Scoreboard JSON string for the current PC2 scoreboard.
+     * @param remoteScoreboardJson a CLICS Scoreboard JSON string for the remote CCS being shadowed.
+     * 
+     * @return an array of comparisons between rows of the PC2 and Remote CCS scoreboards; the returned array
+     *          will be empty (size=0) if an error occurs while processing the input JSON strings.
+     */
+    public ShadowScoreboardRowComparison [] compare (String pc2ScoreboardJson, String remoteScoreboardJson) {
```

Comment:
fyi - this method is nearly 100 lines long and it is not a simple switch or if/else set of blocks.  
In the future consider moving the for rank block of code into its own method.   Then both
that new method and the compare method could have their own unit tests.

Reply:
I am assuming that "fyi" and "In the future" means that you accept that I'm not going to change this now -- for several reasons.  First, there would be strong coupling between such a new method and the existing method, requiring passing multiple parameters and/or changing local variables to being fields (which increases vulnerability).  Second, I'm pretty certain that we all agreed years ago to set the "Warning limit" on method size in PC2 code to 200 lines; this method is less than 50% of that.  Third, I don't think the method is particularly confusing or hard to read as-is:  it starts with a variety of sanity checks and data updates, followed by a main loop which is less than 40 lines (of which over 1/3 are comments/blanks).